generators/chipyard/src/main/scala/TestHarness.scala

```diff
@@ -19,32 +23,96 @@ trait HasTestHarnessFunctions {
 }
 
 trait HasHarnessSignalReferences {
+  // clock/reset of the chiptop reference clock (can be different than the implicit harness clock/reset)
   def harnessClock: Clock
   def harnessReset: Reset
   def dutReset: Reset
   def success: Bool
 }
 
+class HarnessClockInstantiator {
+  private var _clockMap: HashMap[String, (Double, ClockBundle)] = HashMap.empty
+
+  // request a clock bundle at a particular frequency
+  def getClockBundle(name: String, freqRequested: Double): ClockBundle = {
+    val clockBundle = Wire(new ClockBundle(ClockBundleParameters()))
+    _clockMap(name) = (freqRequested, clockBundle)
+    clockBundle
+  }
+
+  // connect all clock wires specified to a divider only PLL
+  def instantiateHarnessDividerPLL(refClock: ClockBundle): Unit = {
+    val sinks = _clockMap.map({ case (name, (freq, bundle)) =>
+      ClockSinkParameters(take=Some(ClockParameters(freqMHz=freq/1000000)),name=Some(name))
+    }).toSeq
+
+    val pllConfig = new SimplePllConfiguration("harnessDividerOnlyClockGenerator", sinks)
+    pllConfig.emitSummaries()
+
+    val dividedClocks = HashMap[Int, Clock]()
+    def instantiateDivider(div: Int): Clock = {
+      val divider = Module(new ClockDividerN(div))
+      divider.suggestName(s"ClockDivideBy${div}")
+      divider.io.clk_in := refClock.clock
+      dividedClocks(div) = divider.io.clk_out
+      divider.io.clk_out
+    }
+
+    // connect wires to clock source
+    for (sinkParams <- sinks) {
+      // bypass the reference freq. (don't create a divider + reset sync)
+      val (divClock, divReset) = if (sinkParams.take.get.freqMHz != pllConfig.referenceFreqMHz) {
+        val div = pllConfig.sinkDividerMap(sinkParams)
+        val divClock = dividedClocks.getOrElse(div, instantiateDivider(div))
+        (divClock, ResetCatchAndSync(divClock, refClock.reset.asBool))
+      } else {
+        (refClock.clock, refClock.reset)
+      }
+
+      _clockMap(sinkParams.name.get)._2.clock := divClock
+      _clockMap(sinkParams.name.get)._2.reset := divReset
+    }
+  }
+}
+
+case object HarnessClockInstantiatorKey extends Field[HarnessClockInstantiator](new HarnessClockInstantiator)
+
 class TestHarness(implicit val p: Parameters) extends Module with HasHarnessSignalReferences {
   val io = IO(new Bundle {
     val success = Output(Bool())
   })
 
+  val harnessClock = Wire(Clock())
+  val harnessReset = Wire(Reset())
+
   val lazyDut = LazyModule(p(BuildTop)(p)).suggestName("chiptop")
-  val dut = Module(lazyDut.module)
+  withClockAndReset(harnessClock, harnessReset) {
+    val dut = Module(lazyDut.module)
```

Comment:
If this is just to avoid having an implicit clock we can change the class to be a `BareModule`.
Otherwise its confusing to me to be passing harnessClock to the dut.

Reply:
Yeah don't do this. The dut already creates a harnessFunction to wire its clock/reset, we don't want to make it seem like this relies on implicit behavior.

Reply:
Huh. Ok. Ill double check and change this.